mongo/jobs.go

```diff
@@ -25,104 +25,81 @@ func (m *JobStore) UnlockJob(ctx context.Context, lockID string) {
 	log.Info(ctx, "job lockID has unlocked successfully")
 }
 
-// CheckNewReindexCanBeCreated checks if a new reindex job can be created depending on if a reindex job is currently in progress between cfg.MaxReindexJobRuntime before now and now
+// CheckInProgressJob checks if a new reindex job can be created depending on if a reindex job is currently in progress between cfg.MaxReindexJobRuntime before now and now
 // It returns an error if a reindex job already exists which is in_progress state
-func (m *JobStore) CheckNewReindexCanBeCreated(ctx context.Context) error {
-	log.Info(ctx, "checking if a new reindex job can be created")
+func (m *JobStore) CheckInProgressJob(ctx context.Context) error {
+	log.Info(ctx, "checking if an existing reindex job is in progress")
 
 	s := m.Session.Copy()
 	defer s.Close()
 
-	// get all the jobs with state "in-progress" and order them by "-reindex_started" starting with most recent
-	iter := s.DB(m.Database).C(m.JobsCollection).Find(bson.M{"state": "in-progress"}).Sort("-reindex_started").Iter()
+	var job models.Job
 
-	// checkFromTime is the time of configured variable "MaxReindexJobRuntime" from now
-	checkFromTime := time.Now().Add(-1 * m.cfg.MaxReindexJobRuntime)
-	result := models.Job{}
-
-	for iter.Next(&result) {
-		jobStartTime := result.ReindexStarted
-
-		// check if start time of the job in progress is later than the checkFromTime but earlier than now
-		if jobStartTime.After(checkFromTime) && jobStartTime.Before(time.Now()) {
-			logData := log.Data{
-				"id":                   result.ID,
-				"state":                result.State,
-				"reindex_started_time": jobStartTime,
-			}
-			log.Info(ctx, "found an existing reindex job in progress", logData)
-			return ErrExistingJobInProgress
+	// get job with state "in-progress" and the most recent "reindex_started" time
+	err := s.DB(m.Database).C(m.JobsCollection).Find(bson.M{"state": "in-progress"}).Sort("-reindex_started").One(&job)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			log.Info(ctx, "no reindex jobs with state in progress")
+			return nil
 		}
+		log.Error(ctx, "failed to get job with state in-progress and the most recent reindex_started time", err)
+		return err
 	}
 
-	defer func() {
-		if err := iter.Close(); err != nil {
-			log.Error(ctx, "error closing iterator", err)
+	// checkFromTime is the time of configured variable "MaxReindexJobRuntime" from now
+	checkFromTime := time.Now().Add(-1 * m.cfg.MaxReindexJobRuntime)
+
+	// check if start time of the job in progress is later than the checkFromTime but earlier than now
+	if job.ReindexStarted.After(checkFromTime) && job.ReindexStarted.Before(time.Now()) {
```

Comment:
I wonder if this could be added to the query to mongo as a filter option?

Reply:
Done